azurerm/resource_arm_mssql_elasticpool.go

```diff
@@ -177,81 +186,188 @@ func resourceArmMsSqlElasticPool() *schema.Resource {
 		CustomizeDiff: func(diff *schema.ResourceDiff, v interface{}) error {
 
 			name, _ := diff.GetOk("sku.0.name")
+			tier, _ := diff.GetOk("sku.0.tier")
 			capacity, _ := diff.GetOk("sku.0.capacity")
+			family, _ := diff.GetOk("sku.0.family")
+			maxSizeBytes, _ := diff.GetOk("max_size_bytes")
+			maxSizeGb, _ := diff.GetOk("max_size_gb")
 			minCapacity, _ := diff.GetOk("per_database_settings.0.min_capacity")
 			maxCapacity, _ := diff.GetOk("per_database_settings.0.max_capacity")
 
-			if strings.HasPrefix(strings.ToLower(name.(string)), "gp_") {
+			// Convert Bytes to Gigabytes
+			if maxSizeBytes != 0 {
+				maxSizeGb = float64(maxSizeBytes.(int) / 1073741824)
+			}
+
+			// Basic Checks
+			if strings.EqualFold(name.(string), "BasicPool") {
+				maxAllowedGB := azure.BasicGetMaxSizeGB(capacity.(int))
+
+				if maxAllowedGB == 0 {
+					return fmt.Errorf("service tier 'Basic' must have a 'capacity'(%d) of 50, 100, 200, 300, 400, 800, 1200 or 1600 DTUs", capacity.(int))
+				}
 
-				if capacity.(int) > 24 {
-					return fmt.Errorf("GeneralPurpose pricing tier only supports upto 24 vCores")
+				// Basic SKU does not let you pick your max_size_GB they are fixed values
+				if maxSizeGb.(float64) != maxAllowedGB {
+					return fmt.Errorf("service tier 'Basic' with a 'capacity' of %d must have a 'max_size_gb' of %.7f GB, got %.7f GB", capacity.(int), maxAllowedGB, maxSizeGb.(float64))
 				}
+			}
+
+			// Standard Checks
+			if strings.EqualFold(name.(string), "StandardPool") {
+				maxAllowedGB := azure.StandardGetMaxSizeGB(capacity.(int))
 
-				if capacity.(int) < 1 {
-					return fmt.Errorf("GeneralPurpose pricing tier must have a minimum of 1 vCores")
+				if maxAllowedGB == 0 {
+					return fmt.Errorf("service tier 'Standard' must have a 'capacity'(%d) of 50, 100, 200, 300, 400, 800, 1200, 1600, 2000, 2500 or 3000 DTUs", capacity.(int))
 				}
 
-				switch {
-				case capacity.(int) == 1:
-				case capacity.(int) == 2:
-				case capacity.(int) == 4:
-				case capacity.(int) == 8:
-				case capacity.(int) == 16:
-				case capacity.(int) == 24:
-				default:
-					return fmt.Errorf("GeneralPurpose pricing tier must have a capacity of 1, 2, 4, 8, 16, or 24 vCores")
+				if maxSizeGb.(float64) > maxAllowedGB {
+					return fmt.Errorf("service tier 'Standard' with a 'capacity' of %d must have a 'max_size_gb' no greater than %d GB, got %d GB", capacity.(int), int(maxAllowedGB), int(maxSizeGb.(float64)))
 				}
 			}
 
-			if strings.HasPrefix(strings.ToLower(name.(string)), "bc_") {
-				if capacity.(int) > 80 {
-					return fmt.Errorf("BusinessCritical pricing tier only supports upto 80 vCores")
+			// Premium Checks
+			if strings.EqualFold(name.(string), "PremiumPool") {
```

Comment:
This is very similar to the standard checks, could we combine these into a function and share some of this?

Reply:
Fixed.